core/src/main/java/org/kiwix/kiwixmobile/core/webserver/ZimHostActivity.kt

```diff
@@ -0,0 +1,290 @@
+        hotspotService = (service as HotspotService.HotspotBinder).service
+        hotspotService!!.registerCallBack(this@ZimHostActivity)
+      }
+
+      override fun onServiceDisconnected(arg0: ComponentName) {}
+    }
+
+    startServerButton.setOnClickListener { startStopServer() }
+  }
+
+  private fun startStopServer() {
+    when {
+      ServerUtils.isServerStarted -> stopServer()
+      selectedBooksPath.size > 0 -> startHotspotManuallyDialog()
+      else -> Toast.makeText(
+        this,
+        R.string.no_books_selected_toast_message,
+        Toast.LENGTH_SHORT
+      ).show()
+    }
+  }
+
+  private fun stopServer() {
+    startService(createHotspotIntent(ACTION_STOP_SERVER))
+  }
+
+  private fun select(bookOnDisk: BooksOnDiskListItem.BookOnDisk) {
+    val booksList: ArrayList<BooksOnDiskListItem> = ArrayList()
+    for (item in booksAdapter.items) {
+      if (item == bookOnDisk) {
+        item.isSelected = !item.isSelected
+      }
+      booksList.add(item)
+    }
+    booksAdapter.items = booksList
+    saveHostedBooks(booksList)
+  }
+
+  override fun onStart() {
+    super.onStart()
+    bindService()
+  }
+
+  override fun onStop() {
+    super.onStop()
+    unbindService()
+  }
+
+  private fun bindService() {
+    bindService(
+      Intent(this, HotspotService::class.java), serviceConnection,
+      Context.BIND_AUTO_CREATE
+    )
+  }
+
+  private fun unbindService() {
+    if (hotspotService != null) {
+      unbindService(serviceConnection)
+      hotspotService!!.registerCallBack(null)
+    }
+  }
+
+  override fun onResume() {
+    super.onResume()
+    presenter.loadBooks(sharedPreferenceUtil.hostedBooks)
+    if (ServerUtils.isServerStarted) {
+      ip = ServerUtils.getSocketAddress()
+      layoutServerStarted()
+    }
+  }
+
+  private fun saveHostedBooks(booksList: ArrayList<BooksOnDiskListItem>) {
+    sharedPreferenceUtil.hostedBooks =
+      booksList.filter { book -> book.isSelected && book is BookOnDisk && book.book.title != null }
+        .map { book -> (book as BookOnDisk).book.title }.toSet()
+  }
+
+  private fun layoutServerStarted() {
+    serverTextView.text = getString(R.string.server_started_message, ip)
+    startServerButton.text = getString(R.string.stop_server_label)
+    startServerButton.setBackgroundColor(resources.getColor(R.color.stopServer))
+    bookDelegate.selectionMode = SelectionMode.NORMAL
+    booksAdapter.notifyDataSetChanged()
+  }
+
+  private fun layoutServerStopped() {
+    serverTextView.text = getString(R.string.server_textview_default_message)
+    startServerButton.text = getString(R.string.start_server_label)
+    startServerButton.setBackgroundColor(resources.getColor(R.color.greenTick))
+    bookDelegate.selectionMode = SelectionMode.MULTI
+    booksAdapter.notifyDataSetChanged()
+  }
+
+  override fun onDestroy() {
+    super.onDestroy()
+    presenter.detachView()
+  }
+
+  private fun setUpToolbar() {
+    val toolbar = findViewById<Toolbar>(R.id.toolbar)
```

Comment:
you shouldn't need findViewById, Kotlin Android Extensions means we can just use the id of the view as the view. You should get an import prompt if you just delete this line

Reply:
https://github.com/kiwix/kiwix-android/pull/1569/commits/38bdabb741fbc163307a82e6d5e93b50dc86ffb1